### DISCUSSION TITLE
chromium 无头模式参数优化

### DISCUSSION THREAD

**Discussion by mark9804** [Arondight/Adachi-BOT#538] [Ideas]
系统：
```
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
```

chromium 版本：

```
$ chromium --version
Chromium 90.0.4430.212 built on Debian bullseye/sid, running on Debian 11.2
```

行为：
![image](https://user-images.githubusercontent.com/9006264/147844076-0df9d3a1-366c-44de-a898-83acb21f2e5e.png)

复现：
1. `npm run stop` 停止机器人，此时负载应当消失；
2. `npm run start` 启动机器人，此时没有 chromium 负载；
3. 执行任意一个需要调用 chromium 的命令；
4. 监视到 Chromium 持续占用 CPU 资源，即使渲染工作已经结束。

**Comment by Arondight** [comment]
确实是这样的，我在代码中用浏览器的逻辑是这样的

1. bot 启动
2. 收到一个绘图命令，阻塞其他绘图命令
3. 直到浏览器启动完成
4. 每个绘图命令使用一个标签进行绘图
5. 绘图完毕关闭标签
6. 继续等待其他绘图命令
7. bot 结束时关闭浏览器

要不然每次都起一个浏览器得等死人

**Reply by mark9804** [reply]
我找到了一个带点魔法的参数集合，不知道会不会稍微好一点
https://stackoverflow.com/a/58589026 (需要删除 `--single-process`)
```javascript
const options = {
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-accelerated-2d-canvas',
      '--no-first-run',
      '--no-zygote',
      '--disable-gpu'
    ],
    headless: true
  }

  return await puppeteer.launch(options)
```

**Reply by Arondight** [reply]
你可以试一下，我的树莓派在公司

**Reply by Arondight** [reply]
不过树莓派记得也有俩大核，应该问题不算大吧

**Reply by mark9804** [reply]
研究了一下，设备是树莓派 4B 4GB，BCM2711，最高 1.8GHz
统一使用 `信息 刻晴` ，从 `render：已生成 genshin-overview 功能的数据调试文件` 开始，到 `开始图片任务` 为止 ：
 - 令chromium 常驻后台的平均渲染时间大约为 2.5s ，chromium 常驻后台消耗内存 120M，CPU 100% （Unix模式计算）
 - chromium 不常驻后台的平均渲染时间大约为 2.6s
 
用的是这个参数

<details>
<summary>完整参数</summary>

```javascript
args: [
          "--no-sandbox",
          "--disable-setuid-sandbox",
          "--disable-translate",
          "--disable-extensions",
          "--disable-sync",
          "--disable-default-apps",
          "--no-first-run",
          "--no-zygote"
        ],
```
</details>

但是实际测试当中，发现 disable 不需要的东西并不能快多少（可能0.0几秒吧，都能算进误差范围了），所以 args 可以精简成


```javascript
args: [
          "--no-sandbox",
          "--disable-setuid-sandbox",
          "--no-first-run",
          "--no-zygote"
        ],
```

在我的测试环境下可以在性能和消耗的计算资源中间获得一个平衡

----

蚌埠住了，多测了几遍发现不 disable 更快

**Comment by mark9804** [comment]
更改参数前后空载状态下对比

改前
![EFCA3031-66D2-411F-9940-5D6E0774BC3B_1_105_c](https://user-images.githubusercontent.com/9006264/147847181-eed1c867-6f83-4fa4-bed7-f99dd34c19b4.jpeg)

----
改后
![585745F4-EADB-4475-A9FA-49C876107A5C_1_102_o](https://user-images.githubusercontent.com/9006264/147847189-c39fcf94-b28d-4831-a13d-dc1ab821ef60.jpeg)

**Reply by Arondight** [reply]
第二张图chrome进程咋没了

**Reply by mark9804** [reply]
> 第二张图chrome进程咋没了

`"--no-zygote"` 参数干掉了，因为使用 `--no-first-run` 跳过了不必要（大概不必要？）的 first run tasks，所以冷启动也能获得差不多的速度

> `--no-first-run`: Skip First Run tasks, whether or not it's actually the First Run. Overridden by kForceFirstRun. This does not drop the First Run sentinel and thus doesn't prevent first run from occuring the next time chrome is launched without this flag.
> `--no-zygote`: Disables the use of a zygote process for forking child processes. Instead, child processes will be forked and exec'd directly. Note that --no-sandbox should also be used together with this flag because the sandbox needs the zygote to work.

[List of Chromium Command Line Switches « Peter Beverloo](https://peter.sh/experiments/chromium-command-line-switches/)

详细的话，如果不加上 `--no-first-run`，那么从生成 debug 文件到截图完成需要 3.6s 左右的时间；
加上 `--no-first-run` 和 `--no-zygote`，从生成 debug 文件到截图完成需要 2.5-2.6s 左右的时间

**Reply by Arondight** [reply]
奇怪，我看这个参数的描述，不会让 chrome 进程消失掉啊 https://github.com/chromium/chromium/blob/main/docs/linux/zygote.md

**Reply by mark9804** [reply]
我还不确定这个是不是有什么奇怪的副作用，自己先用一段时间看看

**Reply by Arondight** [reply]
另外好像我的树莓派3b要比你的快，就算是bot启动后第一次截图，浏览器启动->截图->转换base64，这个过程也只用了半秒，不知道为啥你的会两秒多 https://github.com/Arondight/Adachi-BOT/issues/418#issuecomment-975073788

**Reply by Arondight** [reply]
哦，看错了，这是半分钟

**Reply by Arondight** [reply]
我找到了他们的 gerrit https://codereview.chromium.org/2384163002 ，看描述这个参数也不会让进程消失，只不过是派生进程的方式有了变化，chrome 进程没有了我感觉哪里有些问题

**Reply by mark9804** [reply]
`ps -aef | grep chromium` 能看到 chromium ，可能占用不够高， `top` 看不上眼……

**Reply by Arondight** [reply]
你可以把 viewDebug 打开，看看这些参数到底导致了什么变化，截图之后窗口会被关闭吗，这个参数按照说明必须和 `--no-sandbox` 一起使用，应该和是否工作在无头模式无关

**Reply by mark9804** [reply]
![image](https://user-images.githubusercontent.com/9006264/147847686-a1f0623a-7181-4c9c-828e-856b56951c0b.png)
有进程，CPU消耗 0.0%

**Reply by mark9804** [reply]
> 你可以把 viewDebug 打开，看看这些参数到底导致了什么变化，截图之后窗口会被关闭吗，这个参数按照说明必须和 `--no-sandbox` 一起使用，应该和是否工作在无头模式无关

吃完饭测一下看看

**Reply by Arondight** [reply]
那就不用看了，只要进程还活着那就没问题，我看这个参数的文档肯定是适合我们的使用方式的，直接提上来就好了

**Reply by mark9804** [reply]
> 浏览器启动->截图->转换base64

我觉得浏览器的启动速度应该是和存储卡的读写速度强相关的

我现在用的这张 microSD（TF）大文件性能是 130M/s 读取 90 写入，可能因为这个所以上面 `--disable-…` 这些 flag 测着没什么区别，不然以我朴素的认知来说禁用了用不到的东西应该会更快才对

不过至少追加的这两个 flag 把 4B 风扇动不动就起飞的问题解决了，再去排列组合这几个追求极致性能的 flag 感觉就没什么意义

**Reply by Arondight** [reply]
啊这，我的树莓派3b只有一个裸板，放在老家两年架网站还被海风吹绣了，现在放在公司里连撒热片都掉了，跑点啥我都没关心过它烫不烫的，我感觉这么一比你的树莓派算是非常幸福了
